src/notifications/mod.rs

```diff
@@ -0,0 +1,149 @@
+use std::{env, fmt};
+
+pub const WEBHOOK_URL: &str = "WEBHOOK_URL";
+
+#[derive(Debug, Copy, Clone, Deserialize, Serialize)]
+pub enum NotificationEventType {
+  Broadcast,
+  //
+  Updating,
+  UpdateSuccessful,
+  UpdateFailed,
+  //
+  Starting,
+  StartSuccessful,
+  StartFailed,
+  //
+  Stopping,
+  StopSuccessful,
+  StopFailed,
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct NotificationMessage {
+  event_type: NotificationEventType,
+  event_message: String,
+  timestamp: String,
+}
+
+impl fmt::Display for NotificationEventType {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!(f, "{:?}", self)
+  }
+}
+
+fn fetch_webhook_url() -> Result<String, VarError> {
+  env::var(WEBHOOK_URL)
+}
+
+fn is_webhook_enabled() -> bool {
+  fetch_webhook_url().is_ok()
+}
+
+pub trait NotificationEvent {
+  fn handle_request(&self, request: RequestBuilder);
+  fn create_notification(&self) -> NotificationMessage;
+  fn build_request(&self, webhook_url: &str) -> RequestBuilder;
+  fn send_custom_notification(&self, webhook_url: &str, message: &str);
+  fn send_notification(&self);
+}
+
+impl NotificationEvent for NotificationEventType {
+  fn handle_request(&self, request: RequestBuilder) {
+    let response = request.send();
+    if let Ok(parsed_response) = response {
+      let response_status = parsed_response.status();
+      let response_message = parsed_response.text().unwrap();
+      match response_status.as_u16() {
+        204 | 201 => info!("[{}]: Webhook message sent successfully!", self.to_string()),
+        _ => error!("Request failed! {}, {}", response_status, response_message),
+      }
+    } else {
+      error!(
+        "[{}]: Error with webhook! Status {}",
+        self.to_string(),
+        response
+          .err()
+          .unwrap()
+          .status()
+          .unwrap_or(StatusCode::INTERNAL_SERVER_ERROR)
+          .as_str()
+      );
+    }
+  }
+  fn create_notification(&self) -> NotificationMessage {
+    let name = self.to_string();
+    NotificationMessage {
+      event_type: self.to_owned(),
+      event_message: format!("Server Status: {}", to_title_case(name.as_str())),
+      timestamp: Local::now().format("%m/%d/%Y %H:%M:%S %Z").to_string(),
+    }
+  }
+  fn build_request(&self, webhook_url: &str) -> RequestBuilder {
+    let client = reqwest::blocking::Client::new();
+    client.post(webhook_url)
+  }
+  fn send_custom_notification(&self, webhook_url: &str, message: &str) {
+    let mut notification = self.create_notification();
+    notification.event_message = message.to_string();
+    debug!("Webhook enabled, sending notification {}", self.to_string());
+
+    let mut req = self.build_request(webhook_url);
+    req = if is_discord_webhook(webhook_url) {
+      info!("Sending discord notification <3");
+      req.json(&build_discord_payload(notification))
+    } else {
+      debug!(
+        "Webhook Payload: {}",
+        serde_json::to_string(&notification).unwrap()
+      );
```

Comment:
I tired this, but since a discord json message is widely different than a standard webhook request it would bomb on the type for payload. Do you know how to specify an any type?

Reply:
Oh I didn't even think about that. There is an [`Any` trait](https://doc.rust-lang.org/std/any/trait.Any.html) in rust, but it's not very ergonomic to use, so I think how it is now is fine!